File: frontend/lib/models/invite.dart
import 'package:json_annotation/json_annotation.dart';
import 'note.dart';
import 'user.dart';

part 'invite.g.dart';

enum InviteStatus {
  @JsonValue('pending')
  pending,
  @JsonValue('accepted')
  accepted,
  @JsonValue('declined')
  declined,
}

@JsonSerializable()
class Invite {
  final int id;
  final int noteId;
  final int fromUserId;
  final int toUserId;
  final InviteStatus status;
  final DateTime createdAt;
  final DateTime updatedAt;
  final Note note;
  final User fromUser;
  final User toUser;

  Invite({
    required this.id,
    @JsonKey(name: 'note_id') required this.noteId,
    @JsonKey(name: 'from_user_id') required this.fromUserId,
    @JsonKey(name: 'to_user_id') required this.toUserId,
    required this.status,
    @JsonKey(name: 'created_at') required this.createdAt,
    @JsonKey(name: 'updated_at') required this.updatedAt,
    required this.note,
    @JsonKey(name: 'from_user') required this.fromUser,
    @JsonKey(name: 'to_user') required this.toUser,
  });

  factory Invite.fromJson(Map<String, dynamic> json) => _$InviteFromJson(json);
  Map<String, dynamic> toJson() => _$InviteToJson(this);
}

File: backend/internal/models/invite.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type InviteStatus string

const (
	InviteStatusPending  InviteStatus = "pending"
	InviteStatusAccepted InviteStatus = "accepted"
	InviteStatusDeclined InviteStatus = "declined"
)

type Invite struct {
	ID         uint           `json:"id" gorm:"primaryKey"`
	ThreadID   uint           `json:"thread_id" gorm:"not null"`
	Thread     Thread         `json:"thread" gorm:"foreignKey:ThreadID"`
	FromUserID uint           `json:"from_user_id" gorm:"not null"`
	FromUser   User           `json:"from_user" gorm:"foreignKey:FromUserID"`
	ToUserID   uint           `json:"to_user_id" gorm:"not null"`
	ToUser     User           `json:"to_user" gorm:"foreignKey:ToUserID"`
	Status     InviteStatus   `json:"status" gorm:"default:'pending'"`
	CreatedAt  time.Time      `json:"created_at"`
	UpdatedAt  time.Time      `json:"updated_at"`
	DeletedAt  gorm.DeletedAt `json:"deleted_at,omitempty" gorm:"index"`
}

type CreateInviteRequest struct {
	ThreadID uint `json:"thread_id" binding:"required"`
	ToUserID uint `json:"to_user_id" binding:"required"`
}

type InviteResponse struct {
	ID         uint           `json:"id"`
	ThreadID   uint           `json:"thread_id"`
	Thread     ThreadResponse `json:"thread"`
	FromUserID uint           `json:"from_user_id"`
	FromUser   UserResponse   `json:"from_user"`
	ToUserID   uint           `json:"to_user_id"`
	ToUser     UserResponse   `json:"to_user"`
	Status     InviteStatus   `json:"status"`
	CreatedAt  time.Time      `json:"created_at"`
	UpdatedAt  time.Time      `json:"updated_at"`
}
